File: conformance/adapters/js/adapter.mjs
#!/usr/bin/env node
// Reference engine adapter — the FIRST conforming implementation, using Bahi's engine semantics.
//
//   node adapter.mjs <query> <path-to.khata>   ->   canonical JSON on stdout, exit 0
//
// It uses Node 22's built-in node:sqlite (zero native deps, no npm install), reads the .khata's
// books.sqlite, and answers the language-neutral adapter queries (see ../README.md and
// ../../corpus/manifest.json). Tiers C and R of the harness run through this contract; Folio's
// future Ruby adapter implements the same CLI and must produce byte-identical output.
//
// The audit-chain verifier reuses Bahi's canonicalJson/auditPreimageV2 VERBATIM (canonical-json.mjs)
// — the parity anchor. The report queries are plain SQL over the v12 projection tables.

import { readFileSync, writeFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { randomUUID } from 'node:crypto';
import { DatabaseSync } from 'node:sqlite';
import { readZipEntry } from './zip.mjs';
import { canonicalJson, rowHash, GENESIS_PREV } from './canonical-json.mjs';

// node:sqlite returns INTEGER columns as BigInt; corpus magnitudes are within Number.MAX_SAFE_INTEGER
// (paise sums < 9e15), so coerce to Number for clean integer JSON. Guard against overflow anyway.
function num(v) {
  if (typeof v === 'bigint') {
    if (v > BigInt(Number.MAX_SAFE_INTEGER) || v < BigInt(Number.MIN_SAFE_INTEGER)) {
      throw new Error(`integer ${v} exceeds safe range — report needs BigInt handling`);
    }
    return Number(v);
  }
  return v;
}

function openBooks(khataPath) {
  const buf = readFileSync(khataPath);
  const books = readZipEntry(buf, 'books.sqlite');
  const tmp = join(tmpdir(), `khata-conf-${randomUUID()}.sqlite`);
  writeFileSync(tmp, books);
  const db = new DatabaseSync(tmp, { readOnly: true });
  return { db, tmp };
}

// ─────────────────────────────────────────────────────── Tier C query ─────

function verifyChain(db) {
  const rows = db
    .prepare('SELECT id, ts, actor, action, ref, origin, payload, prev_hash, hash, hash_version FROM audit_log ORDER BY id')
    .all();
  let prev = GENESIS_PREV;
  const badRows = [];
  for (const r of rows) {
    // Chain link: prev_hash must equal the prior row's hash. Genesis (first row) tolerates
    // '' / null / 0*64 per spec §5.2.
    const genesisOk = prev === GENESIS_PREV && (r.prev_hash === '' || r.prev_hash === null || r.prev_hash === GENESIS_PREV);
    if (r.prev_hash !== prev && !genesisOk) {
      badRows.push({ id: num(r.id), reason: 'chain-link', got: r.prev_hash, want: prev });
      prev = r.hash;
      continue;
    }
    const computed = rowHash(r);
    if (computed !== r.hash) {
      badRows.push({ id: num(r.id), reason: 'hash-mismatch', computed, stored: r.hash });
    }
    prev = r.hash;
  }
  return { ok: badRows.length === 0, count: rows.length, badRows: badRows.slice(0, 10) };
}

// ─────────────────────────────────────────────────────── Tier R queries ─────

function trialBalance(db) {
  const rows = db
    .prepare(
      `SELECT el.account_id AS account_id, a.name AS name, a.type AS type,
              SUM(el.debit) AS debit, SUM(el.credit) AS credit
       FROM entry_lines el JOIN accounts a ON a.id = el.account_id
       GROUP BY el.account_id ORDER BY el.account_id`
    )
    .all();
  return rows.map((r) => ({
    account_id: num(r.account_id),
    name: r.name,
    type: r.type,
    debit: num(r.debit),
    credit: num(r.credit),
  }));
}

function accountTypeTotals(db) {
  const rows = db
    .prepare(
      `SELECT a.type AS type, SUM(el.debit) AS debit, SUM(el.credit) AS credit
       FROM entry_lines el JOIN accounts a ON a.id = el.account_id
       GROUP BY a.type ORDER BY a.type`
    )
    .all();
  return rows.map((r) => ({ type: r.type, debit: num(r.debit), credit: num(r.credit) }));
}

function gstOutwardSummary(db) {
  const r = db
    .prepare(
      `SELECT COUNT(*) AS invoice_count,
              COALESCE(SUM(subtotal),0) AS taxable,
              COALESCE(SUM(cgst),0) AS cgst, COALESCE(SUM(sgst),0) AS sgst,
              COALESCE(SUM(igst),0) AS igst, COALESCE(SUM(cess),0) AS cess
       FROM invoices WHERE status = 'posted'`
    )
    .get();
  return {
    invoice_count: num(r.invoice_count),
    taxable: num(r.taxable),
    cgst: num(r.cgst),
    sgst: num(r.sgst),
    igst: num(r.igst),
    cess: num(r.cess),
  };
}

function stockOnHand(db) {
  const rows = db
    .prepare(
      `SELECT sm.item_id AS item_id, COALESCE(i.name,'') AS item_name,
              SUM(CASE WHEN sm.movement_type='in' THEN sm.qty ELSE -sm.qty END) AS on_hand_qty
       FROM stock_movements sm LEFT JOIN items i ON i.id = sm.item_id
       GROUP BY sm.item_id ORDER BY sm.item_id`
    )
    .all();
  return rows.map((r) => ({
    item_id: num(r.item_id),
    item_name: r.item_name,
    on_hand_qty: num(r.on_hand_qty),
  }));
}

const QUERIES = {
  'verify-chain': verifyChain,
  'trial-balance': trialBalance,
  'account-type-totals': accountTypeTotals,
  'gst-outward-summary': gstOutwardSummary,
  'stock-on-hand': stockOnHand,
};

function main() {
  const [query, khataPath] = process.argv.slice(2);
  if (!query || !khataPath || !QUERIES[query]) {
    process.stderr.write(`usage: adapter.mjs <${Object.keys(QUERIES).join('|')}> <file.khata>\n`);
    process.exit(2);
  }
  const { db, tmp } = openBooks(khataPath);
  try {
    const result = QUERIES[query](db);
    process.stdout.write(canonicalJson(result) + '\n');
  } finally {
    db.close();
    rmSync(tmp, { force: true });
  }
}

main();

File: conformance/adapters/ruby/adapter.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Reference engine adapter #2 — Ruby. Proves the conformance contract with a SECOND implementation
# held to the same corpus as the JS adapter, byte-for-byte. This is the shape Folio's Rails engine
# will expose at M1.
#
#   ruby adapter.rb <query> <path-to.khata>   ->   canonical JSON on stdout, exit 0
#
# Reads the .khata's books.sqlite (extracted via the `unzip` CLI) with the sqlite3 gem, and answers
# the language-neutral adapter queries (see ../README.md and ../../corpus/manifest.json). The
# audit-chain verifier reuses the byte contract in canonical_json.rb VERBATIM — the parity anchor.
# Portable across Ruby 2.6 → 3.x.

require 'sqlite3'
require 'tempfile'
require_relative 'canonical_json'

module KhataAdapter
  module_function

  def open_books(khata_path)
    tmp = Tempfile.new(['khata-conf', '.sqlite'])
    tmp.close
    unless system('unzip', '-p', khata_path, 'books.sqlite', out: tmp.path)
      raise "unzip failed to extract books.sqlite from #{khata_path}"
    end
    db = SQLite3::Database.new(tmp.path)
    db.results_as_hash = true
    [db, tmp]
  end

  # ── Tier C — canonicalisation + audit chain ────────────────────────────────
  def verify_chain(db)
    rows = db.execute(
      'SELECT id, ts, actor, action, ref, origin, payload, prev_hash, hash, hash_version ' \
      'FROM audit_log ORDER BY id'
    )
    prev = KhataCanonical::GENESIS_PREV
    bad = []
    rows.each do |r|
      genesis_ok = prev == KhataCanonical::GENESIS_PREV &&
                   (r['prev_hash'] == '' || r['prev_hash'].nil? || r['prev_hash'] == KhataCanonical::GENESIS_PREV)
      if r['prev_hash'] != prev && !genesis_ok
        bad << { 'id' => r['id'], 'reason' => 'chain-link', 'got' => r['prev_hash'], 'want' => prev }
        prev = r['hash']
        next
      end
      computed = KhataCanonical.row_hash(r)
      if computed != r['hash']
        bad << { 'id' => r['id'], 'reason' => 'hash-mismatch', 'computed' => computed, 'stored' => r['hash'] }
      end
      prev = r['hash']
    end
    { 'ok' => bad.empty?, 'count' => rows.length, 'badRows' => bad.first(10) }
  end

  QUERIES = {
    'verify-chain' => method(:verify_chain)
  }.freeze

  def run(argv)
    query, khata_path = argv
    unless query && khata_path && QUERIES.key?(query)
      warn "usage: adapter.rb <#{QUERIES.keys.join('|')}> <file.khata>"
      exit 2
    end
    db, tmp = open_books(khata_path)
    begin
      result = QUERIES[query].call(db)
      $stdout.write(KhataCanonical.canonical_json(result) + "\n")
    ensure
      db.close
      tmp.unlink
    end
  end
end

KhataAdapter.run(ARGV) if $PROGRAM_NAME == __FILE__
